domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,209 @@
+   *
+   * Messages are expected to be resolved serially (though their scheduling can occur across
+   * multiple threads, so order cannot be guaranteed until they're enqueued).
+   */
+  sealed class CommandMessage {
+
+    /** [CommandMessage] for switching the current [AppIconification]. */
+    data class SwitchIconification(val newIconification: AppIconification) : CommandMessage()
+
+    /**
+     * [CommandMessage] that schedules [TakeSnapshot] for the [currentIconification] and [switchId]
+     * to take a [Snapshot] and log-cpu usage after a specific delay period.
+     */
+    data class ScheduleTakeSnapshot(
+      val currentIconification: AppIconification,
+      val switchId: Int
+    ) : CommandMessage()
+
+    /**
+     * [CommandMessage] that takes a real-time [Snapshot] and calls [LogSnapshotDiff] for logging
+     * the relative cpu-usage of the application.
+     *
+     * This relative cpu-usage is calculated by comparing the real-time [Snapshot] with a previous
+     * [Snapshot]. The current [Snapshot] is then put on as the previous [Snapshot] for the next
+     * relative comparison.
+     */
+    data class TakeSnapshot(val switchId: Int) : CommandMessage()
+
+    /** [CommandMessage] for logging the relative cpu usage of the application. */
+    data class LogSnapshotDiff(val relativeCpuUsage: Double) : CommandMessage()
+  }
+
+  /** Returns an appropriate [ScreenName] on the basis of [AppIconification]. */
+  private fun AppIconification.toScreenName(): ScreenName = when (this) {
+    AppIconification.APP_IN_BACKGROUND -> ScreenName.BACKGROUND_SCREEN
+    AppIconification.APP_IN_FOREGROUND -> ScreenName.FOREGROUND_SCREEN
+  }
+
+  /** Returns an appropriate delay time period in millis on the basis of [AppIconification]. */
+  private fun AppIconification.getDelay(): Long = when (this) {
+    AppIconification.APP_IN_BACKGROUND -> backgroundCpuLoggingTimePeriodMillis
+    AppIconification.APP_IN_FOREGROUND -> foregroundCpuLoggingTimePeriodMillis
+  }
+
+  /** Creates an instance of [CpuPerformanceSnapshotter] by properly injecting dependencies. */
+  class Factory @Inject constructor(
+    private val performanceMetricsLogger: PerformanceMetricsLogger,
+    private val consoleLogger: ConsoleLogger,
+    private val exceptionLogger: ExceptionLogger,
+    private val performanceMetricsAssessor: PerformanceMetricsAssessor,
+    @ForegroundCpuLoggingTimePeriodMillis private val foregroundCpuLoggingTimePeriodMillis: Long,
+    @BackgroundCpuLoggingTimePeriodMillis private val backgroundCpuLoggingTimePeriodMillis: Long,
+    @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher
+  ) {
+
+    /** Returns an instance of [CpuPerformanceSnapshotter]. */
+    fun createSnapshotter(): CpuPerformanceSnapshotter =
+      CpuPerformanceSnapshotter(
+        backgroundCoroutineDispatcher,
+        performanceMetricsLogger,
+        AppIconification.APP_IN_FOREGROUND,
```

Comment:
If we think about it, won't the initial state always be foreground? The app won't start without the user starting it and as soon as that happens it'll come to foreground. 
And even if we give this control, how will we ensure that the initial base is set correctly? Because I think, in case of initialising the snapshotter in the lifecycle observer as well, we'll have to give an initial default value for starting this method at some point.

Reply:
To your first point, what if the application is started via WorkManager so that the app can perform its background duties when it's otherwise been shutdown by the system? Won't that be a case where it starts in the background? I actually wonder if we're incorrectly assuming elsewhere that the initial state is foreground--I'm not sure that's actually guaranteed.

To your second point, this goes a bit back to having multiple instances. If we create a new instance with the exact current foreground state, then any changes thereafter can be immediately passed along to the snapshot instance such that there are no gaps and the snapshotter always has the most up-to-date knowledge. Wouldn't that work, or am I missing something?

Reply:
In case of the work manager, I think we had verified this in one of the initial approaches that the app does not start by itself via work manager-- ApplicationStartupListeners weren't activated. 

Also, in any case, if we look at the current implementation, the snapshotter logging really kicks off via updateIconification method and that is present in onAppInForeground and onAppInBackground methods. These methods are the first to be called for notifying us of lifecycle changes and my point being, in any case onAppInForeground is called first (since there's no startup from any external methods like work manager and startup listeners register the lifecycle observer).

Reply:
That actually still sounds wrong to me. Didn't you verify that Application.onCreate is being called after a WorkManager trigger? If you observed that onCreate is called but not the startup listeners, something must be wrong there since the listeners are called directly from onCreate: https://github.com/oppia/oppia-android/blob/20872589968319eecfc8cc5f3426a2f75688c819/app/src/main/java/org/oppia/android/app/application/AbstractOppiaApplication.kt#L48

I think we absolutely need to be sure about the behavior here. The reason I'm concerned is that the application has to start up when opened in the background (otherwise none of our app code will run, OR the app code will run without everything initialized which is pretty seriously broken).

Regarding the updateIconification being the trigger, this might be an oversight. Shouldn't we be initializing logging immediately after creating the snapshotter? Otherwise we lose all CPU activity until a foreground/background activity occurs which might not happen for a while (or at all in some cases). This also seems like a useful case to have a test or two for since it seems like that might be broken in the current implementation.

Reply:
Okay, after further investigation I can say the following things: 
- WorkManager initiation calls `Application.onCreate` and hence it also calls `ApplicationStartupListener.onCreate`. 
- But, these initialisations (which are made from the background when the phone boots) never call our foreground or background methods (which are present in ApplicationLifecycleObserver). 
- These methods are however called when the app is started by the user. But these calls are after the `ApplicationStartupListener.onCreate` call. 

In essence if we: 
a. initialise the snapshotter in `ApplicationStartupListener.onCreate`, it'll always be initialised with **background** Iconification. 
b. initialise multiple snapshotters with `onAppInBackground` or `onAppInForeground` for having the correct Iconification -- here the background initialisation by the WorkManager (for example, after device reboot), won't be able to call these methods and hence, snapshotter won't be initialised at all in this case. 

Inference: 
Apart from `Application.onCreate`, the only other method that is called in all cases (background/foreground startup), is `ApplicationStartupListener.onCreate`. 

I don't think we can make out initial iconification from the onCreate method itself and hence, possible solutions according to me are: 
- Initialise the snapshotter like it's getting initialised right now but with _background_ iconification. 
- Initialise the snapshotter in `ApplicationStartupListener.onCreate`, again with _background_ iconification. 

Other things that I tried: 
- check if work manager is already running during app startup (i.e. from previous initialisation) by getting WorkInfo.state with the help of the TAG that we use. 
- check if a service is already running using activity manager's runningServiceInfo and then finding out if our worker class is something that's in use. 

Both returned false in all cases of app startup and hence weren't usable for us to determine appIconification correctly. 
WDYT?

FYI: I've changed the base iconification to _background_ for now.

Reply:
Hmm so if I'm understanding correctly, we don't get lifecycle callbacks when the app is in the background? To your point on waiting for a stable point to init the snapshotter:
- How does the app behave when going from WorkManager-initiated background to active foreground? Does a new process get created?
- What's the point when "in background" is actually stable? If we don't get lifecycle callbacks, it makes it a bit tricky to know.
- If we default to "in background" how does opening the app via launcher behave? Does the in-foreground callback get called to switch the snapshotter to foreground? Is there an awkward "log milliseconds of CPU time in background" before that point?

Reply:
Yes, we don't get lifecycle callbacks when the app has started in the background. 
Regarding other points: 
- No, the same process gets continued. `onCreate` has already been called so the app just takes it forward from there. 
- It's actually pretty tricky to know this since yes, we don't have lifecycle callbacks. 
- Yes exactly. The `appInForeground` works to update the snapshotter to be in foreground and in almost all cpu logging events I've seen this awkward "log milliseconds of CPU time in background" -- which is possibly due to processing of all `onCreate` methods. It's usually like, the app processes all `onCreate()` and then when it gets stable, it enters the `onForeground` method.

Reply:
Got it. So I suppose with existing signals, we can know sort of retroactively that we're actually in the foreground by analyzing lifecycle callbacks after onCreate() that we'd otherwise not get if the app is background-only, but then the issue is isolating "pre-foreground" CPU time from "background-only" CPU time (though this perspective *does* slightly conflate "background as in the UI is minimized" and "background as in there is no UI").

I spent some time searching, and I don't think there's a reliable way for us to get this info. Perhaps we could delay the initial logging for ~15 seconds to determine if it's in the foreground/background in order to bucket the memory correctly? Or maybe better yet, default to an unknown iconification such that the first logged time window defaults to the latest iconification. Or, would that result in potential inconsistencies when switching between fg/bg early in the app startup?

Reply:
I've tweaked the implementation of the CpuPerformanceSnapshotter per our discussion on chat and have also written additional test cases that cover different [iconification scenarios](https://gist.github.com/BenHenning/6322ebe9b5cff9842080ff103660a7a3).